### DISCUSSION TITLE
NearestNodeDistanceAux not updating correctly

### DISCUSSION THREAD

**Discussion by ghost** [idaholab/moose#25778] [Q&A Modules: Solid mechanics] [answered]
I am running into a situation where NearestNodeDistanceAux will calculate the distance from one surface to another correctly on time = 0, but will not update correctly when only one of the bodies has displacements are defined.

When both blocks have displacements defined on them the distance between the blocks updates throughout the simulation correctly, but when only one has the displacements it calculates the initial distance but doesn't change during the simulation.

Am I missing a setting somewhere to allow it to redo the calc throughout?

**Comment by lindsayad** [comment]
Do you have `use_displaced_mesh = true` in your aux kernel block?

**Reply by lindsayad** [reply]
Actually I see that shouldn't be necessary...
```c++
InputParameters
NearestNodeDistanceAux::validParams()
{
  InputParameters params = AuxKernel::validParams();
  params.addClassDescription(
      "Stores the distance between a block and boundary or between two boundaries.");
  params.addRequiredParam<BoundaryName>("paired_boundary", "The boundary to find the distance to.");
  params.set<bool>("use_displaced_mesh") = true;
  return params;
}
```

**Reply by lindsayad** [reply]
I should have read your post more carefully ... there is likely an issue in the code if this is working with displacements on both blocks but not when you only have displacements on one block. Do you have a minimal example that demonstrates the problem? If not, then we can probably make one ourselves

**Reply by ghost** [reply]
Here is a min working example...if I comment out the explicit displacement lines in the various blocks and turn on the global params version it updates just fine. But when I restrict it to just the middle block or even just two of the three blocks is where it has the issues.

[threeblocks.zip](https://github.com/idaholab/moose/files/13045390/threeblocks.zip)

**Reply by lindsayad** [reply, marked as answer]
Ok so you have to uncomment the `displacements` parameter in the `Mesh` block. The existence of that parameter is what makes us construct the displaced problem. Once you do that and you apply my bug fix in https://github.com/idaholab/moose/pull/25805, then you should see the `gap` results changing over the course of the simulation.

**Reply by ghost** [reply]
Thanks for looking into this. Once that has been pushed I can pull it in and test it out. In the meantime, since this is part of a larger problem using a specific version of moose (from earlier this year) I was looking into something that can be used in the interim until we can update the full package. 

Since I believe the core of the problem that you are stating is that the kernels are creating a coupled problem by default. I noticed that Heat Conduction kernel takes a `displacements` and `use_displaced_mesh` option. If I were to leave the displacments defined in the mesh could I use the `use-displaced_mesh = false` in the kernels blocks to decouple? Would that achieve the same result until we can update the MOOSE install?

**Reply by GiudGiud** [reply]
Hello

That should work. You define displacements globally again and separate the kernels for the coupled and decoupled regions. The decoupled region kernels use `use_displaced_mesh = false`. 

Guillaume

**Reply by ghost** [reply]
Ok that does seem to work now. 

Thanks!

**Comment by GiudGiud** [comment]
Hello

I m assuming he does because it s the default

If we did get the "this is a displaced mesh because there's displacement on at least one block" right here and it s just an update problem, then there is this parameter on the mesh that will let you trigger more updates of the nearest node patches "patch_update_strategy". That could be the issue

if it s not the patches and it s the search, then that section is timed. you can pass `perf_graph_live_time_limit = 0` to the console to see how often it appears

**Reply by ghost** [reply]
It does look like when I have displacements active on all blocks this appears:

```
 Updating Displaced Mesh
          Finished Syncing Displaced Solutions                                           [[33m  0.00 s[39m] [[33m  205 MB[39m]
        Still Updating Displaced Mesh....
          Currently Finding Nearest Nodes..............................                  [[33m  0.00 s[39m] [[33m  205 MB[39m]
          Currently Finding Nearest Nodes...............................                 [[33m  0.00 s[39m] [[33m  205 MB[39m]
        Finished Updating Displaced Mesh                                                 [[33m  0.01 s[39m] [[33m  205 MB[39m]
```
But that does not seem to appear when I only have it on some of the blocks except at the beginning before Time 0.

**Comment by ghost** [comment]
So I ran the `--show-input` command option on the model that does not update the gap distance and it appears that even when passed `use_displaced_mesh = true` it is set to false.

```
 [gap_distance_right]
    type = NearestNodeDistanceAux
    variable = gap
    block = 'middle_block right_block'
    boundary = middle_right
    paired_boundary = right_inside
    use_displaced_mesh = true
[]
```

```
  [./gap_distance_right]
    _unique_action_name          = AddKernelActionadd_aux_kernelAuxKernels/gap_distance_right
    control_tags                 = INVALID
    inactive                     = (no_default)
    isObjectAction               = 1
    type                         = NearestNodeDistanceAux
    block                        = 'middle_block right_block'
    boundary                     = middle_right
    check_boundary_restricted    = 1
    enable                       = 1
    execute_on                   = 'LINEAR TIMESTEP_END'
    paired_boundary              = right_inside
    prop_getter_suffix           = (no_default)
    seed                         = 0
    use_displaced_mesh           = 0
    variable                     = gap
  [../]
[]
```
It also seems to think that the displacements are invalid in the mesh. 
```
[Mesh]
  _unique_action_name            = CreateDisplacedProblemActionMesh
  control_tags                   = (no_default)
  displacements                  = INVALID
  inactive                       = (no_default)
  use_displaced_mesh             = 1
```
My guess is that once the Mesh feels there aren't displacements it will ignore the use of a displaced mesh even though displacement variables are defined on some of the blocks. Is there a way in the Mesh block to tell it there are displacements but only for certain blocks? I would prefer to not have to have displacements turned on for every block as in the much larger model this isn't really necessary and just adds DOFs and computation time that isn't needed.

**Reply by GiudGiud** [reply]
@lindsayad is investigating this.
But yes that seems correct. The displacements are not recognized in one case, causing the use of the undisplaced mesh
